### PR TITLE
Added Element methods and new rules in querySelectorHelper.

### PR DESCRIPTION
**Pull request by Indomian**
- Added support for nesting rule (>, +, ~) in querySelectorHelper.
 - Added nextElementSibling and previousElementSibling to Element class.
 - Added tests for new methods and new querySelector behaviours.

Signed-off-by: Egor Bolgov <e.bolgov@semrush.com>

### CODE REVIEW

File: src/utils/querySelectorHelper.js [hunk 1]
```diff
@@ -46,13 +47,19 @@ export function querySelectorAll(element, query) {
...
  *
- * @param element
- * @param rules
+ * @param {HTMLElement} element - element to check
+ * @param {*} rules - current rules to check against element
+ * @param {boolean} skipSubChildren - flag if true, will not check subchildren of element
  * @return {boolean}
  */
-function* processElement(element, rules) {
+function* processElement(element, rules, skipSubChildren = false) {
```
Comment (christophehurpeau) [inline src/utils/querySelectorHelper.js]: I think I would prefer `processElement` and `processRecursiveElement` or somthing like that over `skipSubChildren`

Comment (christophehurpeau) [inline src/utils/querySelectorHelper.js]: `processChildren` and `processRecursiveChildren` ?

File: src/utils/querySelectorHelper.js [hunk 2]
```diff
@@ -97,7 +104,22 @@ function* processSelectors(element, selectors) {
             return;
         }
 
-        yield* processElement(element, { type: 'ruleSet', rule: rule.rule });
+        switch (rule.rule.nestingOperator) {
+            case '+':
+                if (element.nextElementSibling && matchRule(element.nextElementSibling, rule.rule)) {
```
Comment (christophehurpeau) [inline src/utils/querySelectorHelper.js]: can you keep nextEelemtSibling ?

```js 
const element.nextElementSibling = element.nextElementSibling;
if (nextElementSibling && matchRule(nextElementSibling, rule.rule)) {
    yield nextElementSibling;
}
```

File: tests/src/utils/querySelectorHelper.js [hunk 3]
```diff
@@ -196,3 +196,50 @@ test('Element name equal or starts with "|="', () => {
...
+    let elements = document.body.querySelectorAll('input ~ i');
+    strictEqual(elements.length, 2);
+});
+
+test('Searching for nested structures', () => {
+    const document = new Document();
+    document.body.innerHTML =
+        '<div>' +
+            '<span>' +
+                '<div>' +
```
Comment (christophehurpeau) [inline tests/src/utils/querySelectorHelper.js]: `div` inside `span` should not be allowed... can you use another tag ?

File: src/utils/querySelectorHelper.js (hunk 1 above)
Comment (Indomian) [inline src/utils/querySelectorHelper.js]: May be `processElement` and `processElementDeep` ?

File: src/utils/querySelectorHelper.js (hunk 2 above)
Comment (Indomian) [inline src/utils/querySelectorHelper.js]: Sure, I belive, your variant is faster.

File: tests/src/utils/querySelectorHelper.js (hunk 3 above)
Comment (Indomian) [inline tests/src/utils/querySelectorHelper.js]: No problem, I'll update this.

**Comment by Indomian** [comment]
Rebased over master and code style fixed.

File: src/Element.js [hunk 4]
```diff
@@ -87,6 +87,44 @@ export default class Element extends ParentNode {
...
 
     /**
+     * Returns the {@link Element} immediately following the specified one in its parent's children list,
+     * or null if the specified element is the last one in the list.
+     *
+     * @type {Element}
+     * @readonly
+     */
+    get nextElementSibling() {
+        let children = this.parentNode.children;
```
Comment (christophehurpeau) [inline src/Element.js]: `const siblings = this.parentNode.children` ?

Comment (Indomian) [inline src/Element.js]: Yep... updating.

**Pull request merged**